project build/make/
diff --git a/core/Makefile b/core/Makefile
index b418946..cb767a4 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -2113,8 +2113,7 @@ ifeq (,$(filter true, $(BOARD_USES_FULL_RECOVERY_IMAGE) $(BOARD_USES_RECOVERY_AS
   $(BOARD_BUILD_SYSTEM_ROOT_IMAGE) $(BOARD_INCLUDE_RECOVERY_DTBO) $(BOARD_INCLUDE_RECOVERY_ACPIO) \
   $(BOARD_RAMDISK_USE_LZ4) $(BOARD_RAMDISK_USE_XZ)))
 # Named '.dat' so we don't attempt to use imgdiff for patching it.
-RECOVERY_RESOURCE_ZIP := $(TARGET_OUT_VENDOR)/etc/recovery-resource.dat
-ALL_DEFAULT_INSTALLED_MODULES += $(RECOVERY_RESOURCE_ZIP)
+RECOVERY_RESOURCE_ZIP :=
 else
 RECOVERY_RESOURCE_ZIP :=
 endif
@@ -4998,7 +4997,7 @@ INTERNAL_OTA_PACKAGE_TARGET := $(PRODUCT_OUT)/$(name).zip
 INTERNAL_OTA_METADATA := $(PRODUCT_OUT)/ota_metadata

 ifeq ($(TARGET_BUILD_VARIANT),user)
-    $(INTERNAL_OTA_PACKAGE_TARGET): backuptool := false
+    $(INTERNAL_OTA_PACKAGE_TARGET): backuptool := true
 else
 ifneq ($(LINEAGE_BUILD),)
     $(INTERNAL_OTA_PACKAGE_TARGET): backuptool := true

diff --git a/core/java.mk b/core/java.mk
index a041321..ac665b3 100644
--- a/core/java.mk
+++ b/core/java.mk
@@ -225,7 +225,7 @@ $(full_classes_compiled_jar): PRIVATE_WARNINGS_ENABLE := $(LOCAL_WARNINGS_ENABLE
 # available via JDWP.
 ifneq (,$(PRODUCT_MINIMIZE_JAVA_DEBUG_INFO))
 ifneq (,$(filter userdebug user,$(TARGET_BUILD_VARIANT)))
-LOCAL_JAVACFLAGS+= -g:source,lines
+LOCAL_JAVACFLAGS+= -g:none
 endif
 endif

diff --git a/core/product.mk b/core/product.mk
index 13a182a..32e80d7 100644
--- a/core/product.mk
+++ b/core/product.mk
@@ -560,7 +560,12 @@ _readonly_late_variables += \
   PRODUCT_COPY_FILES \
   PRODUCT_DEX_PREOPT_NEVER_ALLOW_STRIPPING \
   PRODUCT_DEX_PREOPT_BOOT_FLAGS \
-  PRODUCT_SOONG_NAMESPACES
+  PRODUCT_SOONG_NAMESPACES \
+  PRODUCT_DEX_PREOPT_DEFAULT_COMPILER_FILTER \
+  PRODUCT_DEX_PREOPT_DEFAULT_FLAGS \
+  PRODUCT_OTHER_JAVA_DEBUG_INFO \
+  PRODUCT_SYSTEM_SERVER_COMPILER_FILTER \
+  PRODUCT_SYSTEM_SERVER_DEBUG_INFO

 _readonly_early_variables := $(filter-out $(_readonly_late_variables),$(_product_var_list))

diff --git a/target/product/base_system.mk b/target/product/base_system.mk
index 5b118ee..34498b4 100644
--- a/target/product/base_system.mk
+++ b/target/product/base_system.mk
@@ -31,9 +31,7 @@ PRODUCT_PACKAGES += \
     appops \
     app_process \
     appwidget \
-    atrace \
     audioserver \
-    BackupRestoreConfirmation \
     bcc \
     blank_screen \
     blkid \
@@ -44,8 +42,6 @@ PRODUCT_PACKAGES += \
     boringssl_self_test \
     bpfloader \
     bu \
-    bugreport \
-    bugreportz \
     cgroups.json \
     charger \
     cmd \
@@ -68,10 +64,6 @@ PRODUCT_PACKAGES += \
     com.android.wifi \
     ContactsProvider \
     content \
-    crash_dump \
-    CtsShimPrebuilt \
-    CtsShimPrivPrebuilt \
-    debuggerd\
     device_config \
     dmctl \
     dnsmasq \
@@ -79,7 +71,6 @@ PRODUCT_PACKAGES += \
     dpm \
     dumpstate \
     dumpsys \
-    DynamicSystemInstallationService \
     e2fsck \
     ExtShared \
     flags_health_check \
@@ -93,8 +84,6 @@ PRODUCT_PACKAGES += \
     group_system \
     gsid \
     gsi_tool \
-    heapprofd \
-    heapprofd_client \
     gatekeeperd \
     gpuservice \
     hid \
@@ -149,6 +138,7 @@ PRODUCT_PACKAGES += \
     libgui \
     libhardware \
     libhardware_legacy \
+    libincident \
     libinput \
     libinputflinger \
     libiprouteutil \
@@ -256,9 +246,6 @@ PRODUCT_PACKAGES += \
     tc \
     telecom \
     telephony-common \
-    tombstoned \
-    traced \
-    traced_probes \
     tune2fs \
     tzdatacheck \
     uiautomator \
@@ -268,7 +255,6 @@ PRODUCT_PACKAGES += \
     viewcompiler \
     voip-common \
     vold \
-    WallpaperBackup \
     watchdogd \
     wificond \
     wifi.rc \
@@ -356,9 +342,6 @@ endif
 PRODUCT_COPY_FILES += system/core/rootdir/init.zygote32.rc:system/etc/init/hw/init.zygote32.rc
 PRODUCT_DEFAULT_PROPERTY_OVERRIDES += ro.zygote=zygote32

-PRODUCT_SYSTEM_DEFAULT_PROPERTIES += debug.atrace.tags.enableflags=0
-PRODUCT_SYSTEM_DEFAULT_PROPERTIES += persist.traced.enable=1
-
 # Packages included only for eng or userdebug builds, previously debug tagged
 PRODUCT_PACKAGES_DEBUG := \
     adb_keys \
@@ -394,7 +377,6 @@ endif
 # The set of packages whose code can be loaded by the system server.
 PRODUCT_SYSTEM_SERVER_APPS += \
     SettingsProvider \
-    WallpaperBackup

 # Packages included only for eng/userdebug builds, when building with SANITIZE_TARGET=address
 PRODUCT_PACKAGES_DEBUG_ASAN := \

diff --git a/target/product/full_base.mk b/target/product/full_base.mk
index 267859a..31bfedf 100644
--- a/target/product/full_base.mk
+++ b/target/product/full_base.mk
@@ -25,7 +25,6 @@ PRODUCT_PACKAGES := \

 PRODUCT_PACKAGES += \
     LiveWallpapersPicker \
-    PhotoTable \
     preinstalled-packages-platform-full-base.xml

 # Bluetooth:

diff --git a/target/product/handheld_product.mk b/target/product/handheld_product.mk
index 968d150..bffcad7 100644
--- a/target/product/handheld_product.mk
+++ b/target/product/handheld_product.mk
@@ -33,7 +33,6 @@ PRODUCT_PACKAGES += \
     preinstalled-packages-platform-handheld-product.xml \
     QuickSearchBox \
     SettingsIntelligence \
-    frameworks-base-overlays

 ifeq ($(LINEAGE_BUILD),)
 PRODUCT_PACKAGES += \

diff --git a/target/product/handheld_system.mk b/target/product/handheld_system.mk
index f6c92b5..7c71711 100644
--- a/target/product/handheld_system.mk
+++ b/target/product/handheld_system.mk
@@ -32,12 +32,9 @@ $(call inherit-product-if-exists, frameworks/base/data/keyboards/keyboards.mk)
 $(call inherit-product-if-exists, frameworks/webview/chromium/chromium.mk)

 PRODUCT_PACKAGES += \
-    BasicDreams \
     BlockedNumberProvider \
     Bluetooth \
     BluetoothMidiService \
-    BookmarkProvider \
-    BuiltInPrintService \
     CalendarProvider \
     cameraserver \
     CaptivePortalLogin \
@@ -46,7 +43,6 @@ PRODUCT_PACKAGES += \
     clatd.conf \
     DocumentsUI \
     DownloadProviderUi \
-    EasterEgg \
     ExternalStorageProvider \
     FusedLocation \
     InputDevices \
@@ -57,17 +53,13 @@ PRODUCT_PACKAGES += \
     MtpService \
     MusicFX \
     PacProcessor \
-    PrintRecommendationService \
     PrintSpooler \
     ProxyHandler \
     screenrecord \
     SecureElement \
-    SharedStorageBackup \
-    SimAppDialog \
     Telecom \
     TelephonyProvider \
     TeleService \
-    Traceur \
     UserDictionaryProvider \
     VpnDialogs \
     vr \

diff --git a/target/product/handheld_system_ext.mk b/target/product/handheld_system_ext.mk
index d935fbf..8406fee 100644
--- a/target/product/handheld_system_ext.mk
+++ b/target/product/handheld_system_ext.mk
@@ -27,4 +27,3 @@ PRODUCT_PACKAGES += \
     Settings \
     StorageManager \
     SystemUI \
-    WallpaperCropper \

diff --git a/target/product/media_system.mk b/target/product/media_system.mk
index f42491e..ab41631 100644
--- a/target/product/media_system.mk
+++ b/target/product/media_system.mk
@@ -78,10 +78,6 @@ PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
     tombstoned.max_tombstone_count=50
 endif

-PRODUCT_DEFAULT_PROPERTY_OVERRIDES += \
-    ro.logd.size.stats=64K \
-    log.tag.stats_log=I
-
 # Enable CFI for security-sensitive components
 $(call inherit-product, $(SRC_TARGET_DIR)/product/cfi-common.mk)
 $(call inherit-product-if-exists, vendor/google/products/cfi-vendor.mk)

diff --git a/target/product/runtime_libart.mk b/target/product/runtime_libart.mk
index 693b686..0fe5c98 100644
--- a/target/product/runtime_libart.mk
+++ b/target/product/runtime_libart.mk
@@ -61,19 +61,19 @@ ifeq (eng,$(TARGET_BUILD_VARIANT))
         pm.dexopt.boot=extract
 else
     PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
-        pm.dexopt.first-boot=quicken \
-        pm.dexopt.boot=verify
+        pm.dexopt.first-boot=everything \
+        pm.dexopt.boot=everything
 endif

 # The install filter is speed-profile in order to enable the use of
 # profiles from the dex metadata files. Note that if a profile is not provided
 # or if it is empty speed-profile is equivalent to (quicken + empty app image).
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
-    pm.dexopt.install=speed-profile \
-    pm.dexopt.bg-dexopt=speed-profile \
-    pm.dexopt.ab-ota=speed-profile \
-    pm.dexopt.inactive=verify \
-    pm.dexopt.shared=speed
+    pm.dexopt.install=everything \
+    pm.dexopt.bg-dexopt=everything \
+    pm.dexopt.ab-ota=everything \
+    pm.dexopt.inactive=everything \
+    pm.dexopt.shared=everything

 # Pass file with the list of updatable boot class path packages to dex2oat.
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
@@ -87,14 +87,8 @@ PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
     dalvik.vm.dex2oat-max-image-block-size=524288

-# Enable minidebuginfo generation unless overridden.
-PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
-    dalvik.vm.minidebuginfo=true \
-    dalvik.vm.dex2oat-minidebuginfo=true
-
 PRODUCT_USES_DEFAULT_ART_CONFIG := true

 # Disable iorapd by default
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
     ro.iorapd.enable=false
-

diff --git a/target/product/telephony_system.mk b/target/product/telephony_system.mk
index ef48719..4b76ad5 100644
--- a/target/product/telephony_system.mk
+++ b/target/product/telephony_system.mk
@@ -20,8 +20,5 @@
 PRODUCT_PACKAGES := \
     ONS \
     CarrierDefaultApp \
-    CallLogBackup \
-    com.android.cellbroadcast \
-    CellBroadcastLegacyApp \

 PRODUCT_COPY_FILES := \

diff --git a/tools/releasetools/common.py b/tools/releasetools/common.py
index c5758c7..ad459da 100644
--- a/tools/releasetools/common.py
+++ b/tools/releasetools/common.py
@@ -3086,7 +3086,6 @@ def MakeRecoveryPatch(input_dir, output_sink, recovery_img, boot_img,

     d = Difference(recovery_img, boot_img, diff_program=diff_program)
     _, _, patch = d.ComputePatch()
-    output_sink("recovery-from-boot.p", patch)

   try:
     # The following GetTypeAndDevice()s need to use the path in the target
@@ -3141,10 +3140,6 @@ fi
        'recovery_device': recovery_device + '$(getprop ro.boot.slot_suffix)',
        'bonus_args': bonus_args}

-  # The install script location moved from /system/etc to /system/bin in the L
-  # release. In the R release it is in VENDOR/bin or SYSTEM/vendor/bin.
-  output_sink("bin/install-recovery.sh", sh.encode())
-

 class DynamicPartitionUpdate(object):
   def __init__(self, src_group=None, tgt_group=None, progress=None,

diff --git a/tools/releasetools/ota_from_target_files.py b/tools/releasetools/ota_from_target_files.py
index 479a898..63bf826 100755
--- a/tools/releasetools/ota_from_target_files.py
+++ b/tools/releasetools/ota_from_target_files.py
@@ -758,8 +758,6 @@ def WriteFullOTAPackage(input_zip, output_file):
       metadata=metadata,
       info_dict=OPTIONS.info_dict)

-  assert HasRecoveryPatch(input_zip, info_dict=OPTIONS.info_dict)
-
   # Assertions (e.g. downgrade check, device properties check).
   #ts = target_info.GetBuildProp("ro.build.date.utc")
   #ts_text = target_info.GetBuildProp("ro.build.date")

project external/bash/
diff --git a/Android.mk b/Android.mk
deleted file mode 100644
index 69b2a11..0000000
--- a/Android.mk
+++ /dev/null

project external/nano/
diff --git a/Android.mk b/Android.mk
deleted file mode 100644
index 4fcd436..0000000
--- a/Android.mk
+++ /dev/null

project external/noto-fonts/
diff --git a/cjk/Android.bp b/cjk/Android.bp
index 5ff7204..77d4f0b 100644
--- a/cjk/Android.bp
+++ b/cjk/Android.bp
@@ -13,8 +13,13 @@
 // limitations under the License.

 prebuilt_font {
-    name: "NotoSansCJK-Regular.ttc",
-    src: "NotoSansCJK-Regular.ttc",
+    name: "NotoSansCJK.ttc",
+    src: "NotoSansCJK.ttc",
+}
+
+prebuilt_font {
+    name: "NotoSerifCJK-Bold.ttc",
+    src: "NotoSerifCJK-Bold.ttc",
 }

 prebuilt_font {

diff --git a/cjk/NotoSansCJK-Regular.ttc b/cjk/NotoSansCJK-Regular.ttc
index 7fc3586..a2033d0 100644
Binary files a/cjk/NotoSansCJK-Regular.ttc and b/cjk/NotoSansCJK-Regular.ttc differ

diff --git a/cjk/NotoSerifCJK-Regular.ttc b/cjk/NotoSerifCJK-Regular.ttc
index f4ee450..3269d9e 100644
Binary files a/cjk/NotoSerifCJK-Regular.ttc and b/cjk/NotoSerifCJK-Regular.ttc differ

diff --git a/cjk/subset_noto_cjk.py b/cjk/subset_noto_cjk.py
index 5a324bf..be4b8fa 100755
--- a/cjk/subset_noto_cjk.py
+++ b/cjk/subset_noto_cjk.py
@@ -100,4 +100,6 @@ def remove_codepoints_from_ttc(ttc_name):


 remove_codepoints_from_ttc('NotoSansCJK-Regular.ttc')
+remove_codepoints_from_ttc('NotoSansCJK.ttc')
+remove_codepoints_from_ttc('NotoSerifCJK-Bold.ttc')
 remove_codepoints_from_ttc('NotoSerifCJK-Regular.ttc')

diff --git a/fonts.mk b/fonts.mk
index 13bfbfb..c5d3e29 100644
--- a/fonts.mk
+++ b/fonts.mk
@@ -49,7 +49,7 @@ PRODUCT_PACKAGES := \
     NotoSansCham-Bold.ttf \
     NotoSansCham-Regular.ttf \
     NotoSansCherokee-Regular.ttf \
-    NotoSansCJK-Regular.ttc \
+    NotoSansCJK.ttc \
     NotoSansCoptic-Regular.ttf \
     NotoSansCuneiform-Regular.ttf \
     NotoSansCypriot-Regular.ttf \
@@ -209,6 +209,7 @@ PRODUCT_PACKAGES := \
     NotoSerifArmenian-Regular.otf \
     NotoSerifBengali-Bold.ttf \
     NotoSerifBengali-Regular.ttf \
+    NotoSerifCJK-Bold.ttc \
     NotoSerifCJK-Regular.ttc \
     NotoSerifDevanagari-Bold.ttf \
     NotoSerifDevanagari-Regular.ttf \

project external/vim/
diff --git a/Android.mk b/Android.mk
deleted file mode 100644
index c7e2f3c..0000000
--- a/Android.mk
+++ /dev/null

project frameworks/base/
diff --git a/api/current.txt b/api/current.txt
index 952ccdad..1c199adb 100644
--- a/api/current.txt
+++ b/api/current.txt
@@ -77,6 +77,7 @@ package android {
     field public static final String DIAGNOSTIC = "android.permission.DIAGNOSTIC";
     field public static final String DISABLE_KEYGUARD = "android.permission.DISABLE_KEYGUARD";
     field public static final String DUMP = "android.permission.DUMP";
+    field public static final String FAKE_PACKAGE_SIGNATURE = "android.permission.FAKE_PACKAGE_SIGNATURE";
     field public static final String EXPAND_STATUS_BAR = "android.permission.EXPAND_STATUS_BAR";
     field public static final String FACTORY_TEST = "android.permission.FACTORY_TEST";
     field public static final String FOREGROUND_SERVICE = "android.permission.FOREGROUND_SERVICE";
@@ -182,6 +183,7 @@ package android {
     field public static final String CALL_LOG = "android.permission-group.CALL_LOG";
     field public static final String CAMERA = "android.permission-group.CAMERA";
     field public static final String CONTACTS = "android.permission-group.CONTACTS";
+    field public static final String FAKE_PACKAGE = "android.permission-group.FAKE_PACKAGE";
     field public static final String LOCATION = "android.permission-group.LOCATION";
     field public static final String MICROPHONE = "android.permission-group.MICROPHONE";
     field public static final String PHONE = "android.permission-group.PHONE";
@@ -82224,4 +82226,3 @@ package org.xmlpull.v1.sax2 {
   }

 }
-

diff --git a/core/res/AndroidManifest.xml b/core/res/AndroidManifest.xml
index 9bb5aecd..43c095a7 100644
--- a/core/res/AndroidManifest.xml
+++ b/core/res/AndroidManifest.xml
@@ -2853,6 +2853,20 @@
         android:description="@string/permdesc_getPackageSize"
         android:protectionLevel="normal" />

+    <!-- Dummy user-facing group for faking package signature -->
+    <permission-group android:name="android.permission-group.FAKE_PACKAGE"
+        android:label="@string/permgrouplab_fake_package_signature"
+        android:description="@string/permgroupdesc_fake_package_signature"
+        android:request="@string/permgrouprequest_fake_package_signature"
+        android:priority="100" />
+
+    <!-- Allows an application to change the package signature as seen by applications -->
+    <permission android:name="android.permission.FAKE_PACKAGE_SIGNATURE"
+        android:permissionGroup="android.permission-group.UNDEFINED"
+        android:protectionLevel="dangerous"
+        android:label="@string/permlab_fakePackageSignature"
+        android:description="@string/permdesc_fakePackageSignature" />
+
     <!-- @deprecated No longer useful, see
          {@link android.content.pm.PackageManager#addPackageToPreferred}
          for details. -->

diff --git a/core/res/res/values-zh-rCN/strings.xml b/core/res/res/values-zh-rCN/strings.xml
index 74dde0ec..b1c6fb4c 100644
--- a/core/res/res/values-zh-rCN/strings.xml
+++ b/core/res/res/values-zh-rCN/strings.xml
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
+<!--
 /* //device/apps/common/assets/res/any/strings.xml
 **
 ** Copyright 2006, The Android Open Source Project
@@ -2177,6 +2177,11 @@
     <string name="PERSOSUBSTATE_SIM_ICCID_SUCCESS" msgid="8058678548991999545">"ICCID 解锁成功。"</string>
     <string name="PERSOSUBSTATE_SIM_IMPI_SUCCESS" msgid="2545608067978550571">"IMPI 解锁成功。"</string>
     <string name="PERSOSUBSTATE_SIM_NS_SP_SUCCESS" msgid="4352382949744625007">"网络子集服务提供商解锁成功。"</string>
+    <string name="permlab_fakePackageSignature">模拟软件包签名</string>
+    <string name="permdesc_fakePackageSignature">允许当前应用伪装成为其他应用</string>
+    <string name="permgrouplab_fake_package_signature">模拟软件包签名</string>
+    <string name="permgroupdesc_fake_package_signature">伪装成为其他应用</string>
+    <string name="permgrouprequest_fake_package_signature">允许 &lt;b><xliff:g id="app_name" example="Gmail">%1$s</xliff:g>&lt;/b> 模拟软件包签名？</string>
     <string name="config_pdp_reject_dialog_title" msgid="4072057179246785727"></string>
     <string name="config_pdp_reject_user_authentication_failed" msgid="4531693033885744689"></string>
     <string name="config_pdp_reject_service_not_subscribed" msgid="8190338397128671588"></string>

diff --git a/core/res/res/values-zh-rHK/strings.xml b/core/res/res/values-zh-rHK/strings.xml
index 6cae6439..458fefce 100644
--- a/core/res/res/values-zh-rHK/strings.xml
+++ b/core/res/res/values-zh-rHK/strings.xml
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
+<!--
 /* //device/apps/common/assets/res/any/strings.xml
 **
 ** Copyright 2006, The Android Open Source Project
@@ -2177,6 +2177,11 @@
     <string name="PERSOSUBSTATE_SIM_ICCID_SUCCESS" msgid="8058678548991999545">"ICCID 解鎖成功。"</string>
     <string name="PERSOSUBSTATE_SIM_IMPI_SUCCESS" msgid="2545608067978550571">"IMPI 解鎖成功。"</string>
     <string name="PERSOSUBSTATE_SIM_NS_SP_SUCCESS" msgid="4352382949744625007">"網絡子集服務供應商解鎖成功。"</string>
+    <string name="permlab_fakePackageSignature">僞造應用程式封裝簽章</string>
+    <string name="permdesc_fakePackageSignature">允許當前應用程式偽裝成為其他應用程式</string>
+    <string name="permgrouplab_fake_package_signature">僞造應用程式封裝簽章</string>
+    <string name="permgroupdesc_fake_package_signature">偽裝成為其他應用程式</string>
+    <string name="permgrouprequest_fake_package_signature">允許 &lt;b><xliff:g id="app_name" example="Gmail">%1$s</xliff:g>&lt;/b> 僞造應用程式封裝簽章？</string>
     <string name="config_pdp_reject_dialog_title" msgid="4072057179246785727"></string>
     <string name="config_pdp_reject_user_authentication_failed" msgid="4531693033885744689"></string>
     <string name="config_pdp_reject_service_not_subscribed" msgid="8190338397128671588"></string>

diff --git a/core/res/res/values-zh-rTW/strings.xml b/core/res/res/values-zh-rTW/strings.xml
index 58af46e6..69d9a101 100644
--- a/core/res/res/values-zh-rTW/strings.xml
+++ b/core/res/res/values-zh-rTW/strings.xml
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
+<!--
 /* //device/apps/common/assets/res/any/strings.xml
 **
 ** Copyright 2006, The Android Open Source Project
@@ -2177,6 +2177,11 @@
     <string name="PERSOSUBSTATE_SIM_ICCID_SUCCESS" msgid="8058678548991999545">"ICCID 解鎖成功。"</string>
     <string name="PERSOSUBSTATE_SIM_IMPI_SUCCESS" msgid="2545608067978550571">"IMPI 解鎖成功。"</string>
     <string name="PERSOSUBSTATE_SIM_NS_SP_SUCCESS" msgid="4352382949744625007">"網路子集服務供應商解鎖成功。"</string>
+    <string name="permlab_fakePackageSignature">僞造應用程式封裝簽章</string>
+    <string name="permdesc_fakePackageSignature">允許當前應用程式偽裝成為其他應用程式</string>
+    <string name="permgrouplab_fake_package_signature">僞造應用程式封裝簽章</string>
+    <string name="permgroupdesc_fake_package_signature">偽裝成為其他應用程式</string>
+    <string name="permgrouprequest_fake_package_signature">允許 &lt;b><xliff:g id="app_name" example="Gmail">%1$s</xliff:g>&lt;/b> 僞造應用程式封裝簽章？</string>
     <string name="config_pdp_reject_dialog_title" msgid="4072057179246785727"></string>
     <string name="config_pdp_reject_user_authentication_failed" msgid="4531693033885744689"></string>
     <string name="config_pdp_reject_service_not_subscribed" msgid="8190338397128671588"></string>

diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index f4efcc7e..51b461e7 100644
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -1654,6 +1654,8 @@
     <string-array name="config_locationProviderPackageNames" translatable="false">
         <!-- The standard AOSP fused location provider -->
         <item>com.android.location.fused</item>
+        <!-- Google Play Services or microG (free reimplementation) location provider -->
+        <item>com.google.android.gms</item>
     </string-array>

     <!-- This string array can be overriden to enable test location providers initially. -->

diff --git a/core/res/res/values/strings.xml b/core/res/res/values/strings.xml
index 5c659123..c2ab0d54 100644
--- a/core/res/res/values/strings.xml
+++ b/core/res/res/values/strings.xml
@@ -847,6 +847,17 @@

     <!--  Permissions -->

+    <!-- Title of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
+    <string name="permlab_fakePackageSignature">Spoof package signature</string>
+    <!-- Description of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
+    <string name="permdesc_fakePackageSignature">Allows the app to pretend to be a different app.</string>
+    <!-- Title of a category of application permissions, listed so the user can choose whether they want to allow the application to do this. -->
+    <string name="permgrouplab_fake_package_signature">Spoof package signature</string>
+    <!-- Description of a category of application permissions, listed so the user can choose whether they want to allow the application to do this. -->
+    <string name="permgroupdesc_fake_package_signature">pretend to be a different app</string>
+    <!-- Message shown to the user when the apps requests permission from this group. If ever possible this should stay below 80 characters (assuming the parameters takes 20 characters). Don't abbreviate until the message reaches 120 characters though. [CHAR LIMIT=120] -->
+    <string name="permgrouprequest_fake_package_signature">Allow &lt;b><xliff:g id="app_name" example="Gmail">%1$s</xliff:g>&lt;/b> to spoof package signature?</string>
+
     <!-- Title of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
     <string name="permlab_statusBar">disable or modify status bar</string>
     <!-- Description of an application permission, listed so the user can choose whether they want to allow the application to do this. -->

diff --git a/data/fonts/fonts.xml b/data/fonts/fonts.xml
index a678f715..a3463c48 100644
--- a/data/fonts/fonts.xml
+++ b/data/fonts/fonts.xml
@@ -576,20 +576,48 @@
         <font weight="400" style="normal">NotoSansSymbols-Regular-Subsetted.ttf</font>
     </family>
     <family lang="zh-Hans">
-        <font weight="400" style="normal" index="2">NotoSansCJK-Regular.ttc</font>
+        <font weight="100" style="normal" index="2">NotoSansCJK.ttc</font>
+        <font weight="300" style="normal" index="7">NotoSansCJK.ttc</font>
+        <font weight="350" style="normal" index="12">NotoSansCJK.ttc</font>
+        <font weight="400" style="normal" index="27">NotoSansCJK.ttc</font>
+        <font weight="500" style="normal" index="17">NotoSansCJK.ttc</font>
+        <font weight="700" style="normal" index="37">NotoSansCJK.ttc</font>
+        <font weight="900" style="normal" index="22">NotoSansCJK.ttc</font>
         <font weight="400" style="normal" index="2" fallbackFor="serif">NotoSerifCJK-Regular.ttc</font>
+        <font weight="700" style="normal" index="2" fallbackFor="serif">NotoSerifCJK-Bold.ttc</font>
     </family>
     <family lang="zh-Hant,zh-Bopo">
-        <font weight="400" style="normal" index="3">NotoSansCJK-Regular.ttc</font>
+        <font weight="100" style="normal" index="3">NotoSansCJK.ttc</font>
+        <font weight="300" style="normal" index="8">NotoSansCJK.ttc</font>
+        <font weight="350" style="normal" index="13">NotoSansCJK.ttc</font>
+        <font weight="400" style="normal" index="28">NotoSansCJK.ttc</font>
+        <font weight="500" style="normal" index="18">NotoSansCJK.ttc</font>
+        <font weight="700" style="normal" index="38">NotoSansCJK.ttc</font>
+        <font weight="900" style="normal" index="23">NotoSansCJK.ttc</font>
         <font weight="400" style="normal" index="3" fallbackFor="serif">NotoSerifCJK-Regular.ttc</font>
+        <font weight="700" style="normal" index="3" fallbackFor="serif">NotoSerifCJK-Bold.ttc</font>
     </family>
     <family lang="ja">
-        <font weight="400" style="normal" index="0">NotoSansCJK-Regular.ttc</font>
+        <font weight="100" style="normal" index="0">NotoSansCJK.ttc</font>
+        <font weight="300" style="normal" index="5">NotoSansCJK.ttc</font>
+        <font weight="350" style="normal" index="10">NotoSansCJK.ttc</font>
+        <font weight="400" style="normal" index="25">NotoSansCJK.ttc</font>
+        <font weight="500" style="normal" index="15">NotoSansCJK.ttc</font>
+        <font weight="700" style="normal" index="35">NotoSansCJK.ttc</font>
+        <font weight="900" style="normal" index="20">NotoSansCJK.ttc</font>
         <font weight="400" style="normal" index="0" fallbackFor="serif">NotoSerifCJK-Regular.ttc</font>
+        <font weight="700" style="normal" index="0" fallbackFor="serif">NotoSerifCJK-Bold.ttc</font>
     </family>
     <family lang="ko">
-        <font weight="400" style="normal" index="1">NotoSansCJK-Regular.ttc</font>
+        <font weight="100" style="normal" index="1">NotoSansCJK.ttc</font>
+        <font weight="300" style="normal" index="6">NotoSansCJK.ttc</font>
+        <font weight="350" style="normal" index="11">NotoSansCJK.ttc</font>
+        <font weight="400" style="normal" index="26">NotoSansCJK.ttc</font>
+        <font weight="500" style="normal" index="16">NotoSansCJK.ttc</font>
+        <font weight="700" style="normal" index="36">NotoSansCJK.ttc</font>
+        <font weight="900" style="normal" index="21">NotoSansCJK.ttc</font>
         <font weight="400" style="normal" index="1" fallbackFor="serif">NotoSerifCJK-Regular.ttc</font>
+        <font weight="700" style="normal" index="1" fallbackFor="serif">NotoSerifCJK-Bold.ttc</font>
     </family>
     <family lang="und-Zsye">
         <font weight="400" style="normal">NotoColorEmoji.ttf</font>

diff --git a/non-updatable-api/current.txt b/non-updatable-api/current.txt
index 5f15216e..6851cf14 100644
--- a/non-updatable-api/current.txt
+++ b/non-updatable-api/current.txt
@@ -79,6 +79,7 @@ package android {
     field public static final String DUMP = "android.permission.DUMP";
     field public static final String EXPAND_STATUS_BAR = "android.permission.EXPAND_STATUS_BAR";
     field public static final String FACTORY_TEST = "android.permission.FACTORY_TEST";
+    field public static final String FAKE_PACKAGE_SIGNATURE = "android.permission.FAKE_PACKAGE_SIGNATURE";
     field public static final String FOREGROUND_SERVICE = "android.permission.FOREGROUND_SERVICE";
     field public static final String GET_ACCOUNTS = "android.permission.GET_ACCOUNTS";
     field public static final String GET_ACCOUNTS_PRIVILEGED = "android.permission.GET_ACCOUNTS_PRIVILEGED";
@@ -182,6 +183,7 @@ package android {
     field public static final String CALL_LOG = "android.permission-group.CALL_LOG";
     field public static final String CAMERA = "android.permission-group.CAMERA";
     field public static final String CONTACTS = "android.permission-group.CONTACTS";
+    field public static final String FAKE_PACKAGE = "android.permission-group.FAKE_PACKAGE";
     field public static final String LOCATION = "android.permission-group.LOCATION";
     field public static final String MICROPHONE = "android.permission-group.MICROPHONE";
     field public static final String PHONE = "android.permission-group.PHONE";
@@ -80385,4 +80387,3 @@ package org.xmlpull.v1.sax2 {
   }

 }
-

diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index ca9d1c6f..a562b918 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -4465,8 +4465,8 @@ public class PackageManagerService extends IPackageManager.Stub
                 });
             }

-            PackageInfo packageInfo = PackageInfoUtils.generate(p, gids, flags,
-                    ps.firstInstallTime, ps.lastUpdateTime, permissions, state, userId, ps);
+            PackageInfo packageInfo = mayFakeSignature(p, PackageInfoUtils.generate(p, gids, flags,
+                    ps.firstInstallTime, ps.lastUpdateTime, permissions, state, userId, ps), permissions);

             if (packageInfo == null) {
                 return null;
@@ -4502,6 +4502,23 @@ public class PackageManagerService extends IPackageManager.Stub
         }
     }

+    private PackageInfo mayFakeSignature(AndroidPackage p, PackageInfo pi, Set<String> permissions) {
+        try {
+            if (permissions.contains("android.permission.FAKE_PACKAGE_SIGNATURE")
+                    && p.getTargetSdkVersion() > Build.VERSION_CODES.LOLLIPOP_MR1
+                    && p.getMetaData() != null) {
+                String sig = p.getMetaData().getString("fake-signature");
+                if (sig != null) {
+                    pi.signatures = new Signature[] {new Signature(sig)};
+                }
+            }
+        } catch (Throwable t) {
+            // We should never die because of any failures, this is system code!
+            Log.w("PackageManagerService.FAKE_PACKAGE_SIGNATURE", t);
+        }
+        return pi;
+    }
+
     @Override
     public void checkPackageStartable(String packageName, int userId) {
         final int callingUid = Binder.getCallingUid();

project frameworks/native/
diff --git a/cmds/atrace/Android.bp b/cmds/atrace/Android.bp
deleted file mode 100644
index e7d0ad0..0000000
--- a/cmds/atrace/Android.bp
+++ /dev/null

project packages/apps/LineageParts/
diff --git a/res/values-zh-rCN/strings.xml b/res/values-zh-rCN/strings.xml
index 1a5ff77..0ba8acd 100644
--- a/res/values-zh-rCN/strings.xml
+++ b/res/values-zh-rCN/strings.xml
@@ -181,6 +181,8 @@
     <string name="home_answer_call_summary">按 Home 键接听来电</string>
     <string name="extras_title">附加功能</string>
     <string name="additional_buttons_title">附加按钮</string>
+    <string name="edge_gesture_title">边缘触控</string>
+    <string name="edge_gesture_summary">双击手机侧边可触发返回操作</string>
     <string name="click_partial_screenshot_title">点击以进行局部截图</string>
     <string name="click_partial_screenshot_summary">同时按音量下和电源键以截取局部截图</string>
     <string name="button_backlight_title">背光</string>
@@ -463,6 +465,8 @@
     <string name="auto_power_save_summary_on">电量 %s 时自动启用省电</string>
     <string name="auto_power_save_summary_off">不要自动启用省电</string>
     <string name="auto_power_save_never">永不</string>
+    <string name="cpu_overclock_title">高性能模式</string>
+    <string name="cpu_overclock_summary">提升处理器最大功率以改善性能，但可能缩短续航时间</string>
     <string name="long_screen_settings_title">全屏应用</string>
     <string name="long_screen_settings_summary">强制旧式应用程序使用全屏长宽比</string>
     <string name="long_screen_settings_no_apps">无应用</string>

diff --git a/res/values-zh-rTW/strings.xml b/res/values-zh-rTW/strings.xml
index 3a12ffa..c1fd160 100644
--- a/res/values-zh-rTW/strings.xml
+++ b/res/values-zh-rTW/strings.xml
@@ -181,6 +181,8 @@
     <string name="home_answer_call_summary">按主畫面鍵接聽來電</string>
     <string name="extras_title">附加功能</string>
     <string name="additional_buttons_title">附加按鍵</string>
+    <string name="edge_gesture_title">邊緣觸控</string>
+    <string name="edge_gesture_summary">雙擊手機側面以返回</string>
     <string name="click_partial_screenshot_title">點擊以局部螢幕截圖</string>
     <string name="click_partial_screenshot_summary">短按音量下鍵與電源鍵以局部螢幕截圖</string>
     <string name="button_backlight_title">背光</string>
@@ -461,6 +463,8 @@
     <string name="auto_power_save_summary_on">在 %s 電量時自動啟用省電模式</string>
     <string name="auto_power_save_summary_off">不要自動啟用省電模式</string>
     <string name="auto_power_save_never">永不</string>
+    <string name="cpu_overclock_title">高效能模式</string>
+    <string name="cpu_overclock_summary">提高處理器功耗上限以增進性能，但可能縮短續航時間</string>
     <string name="long_screen_settings_title">全螢幕應用程式</string>
     <string name="long_screen_settings_summary">強制舊型應用程式使用全螢幕的比例</string>
     <string name="long_screen_settings_no_apps">沒有應用程式</string>

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 8fe201e..afbde60 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -215,6 +215,8 @@
     <string name="home_answer_call_summary">Answer incoming calls by pressing the home button</string>
     <string name="extras_title">Extras</string>
     <string name="additional_buttons_title">Additional buttons</string>
+    <string name="edge_gesture_title">Edge gesture</string>
+    <string name="edge_gesture_summary">Double tap the phone\'s edge causes back</string>
     <string name="click_partial_screenshot_title">Click to partial screenshot</string>
     <string name="click_partial_screenshot_summary">Short click Volume Down and Power to take partial screenshot</string>

@@ -596,6 +598,8 @@
     <string name="auto_power_save_summary_on">Automatically enable power save mode at %s battery</string>
     <string name="auto_power_save_summary_off">Do not enable power save mode automatically</string>
     <string name="auto_power_save_never">Never</string>
+    <string name="cpu_overclock_title">High performance mode</string>
+    <string name="cpu_overclock_summary">Increase maximum CPU power to improve performance, at the cost of battery life</string>

     <!-- Applications: Long screen -->
     <string name="long_screen_settings_title">Full screen apps</string>

diff --git a/res/xml/button_settings.xml b/res/xml/button_settings.xml
index 5234a96..236947f 100644
--- a/res/xml/button_settings.xml
+++ b/res/xml/button_settings.xml
@@ -336,6 +336,12 @@
                 android:action="org.lineageos.settings.device.ADDITIONAL_BUTTONS_SETTINGS" />
         </lineageos.preference.RemotePreference>

+        <SwitchPreference
+            android:key="edge_gesture"
+            android:title="@string/edge_gesture_title"
+            android:summary="@string/edge_gesture_summary"
+            android:persistent="false" />
+
         <lineageos.preference.LineageSystemSettingSwitchPreference
             android:key="click_partial_screenshot"
             android:title="@string/click_partial_screenshot_title"

diff --git a/res/xml/perf_profile_settings.xml b/res/xml/perf_profile_settings.xml
index 0b301a9..7a129d5 100644
--- a/res/xml/perf_profile_settings.xml
+++ b/res/xml/perf_profile_settings.xml
@@ -37,6 +37,18 @@

     </PreferenceCategory>

+    <PreferenceCategory
+        android:key="cpu_overclock_category"
+        android:title="@string/advanced">
+
+        <SwitchPreference
+            android:key="cpu_overclock"
+            android:title="@string/cpu_overclock_title"
+            android:summary="@string/cpu_overclock_summary"
+            android:persistent="false" />
+
+    </PreferenceCategory>
+
     <PreferenceCategory
         android:key="perf_profile_category"
         android:title="@string/perf_profile_category_title">

diff --git a/src/org/lineageos/lineageparts/input/ButtonSettings.java b/src/org/lineageos/lineageparts/input/ButtonSettings.java
index 613e847..28455bd 100644
--- a/src/org/lineageos/lineageparts/input/ButtonSettings.java
+++ b/src/org/lineageos/lineageparts/input/ButtonSettings.java
@@ -29,6 +29,7 @@ import android.content.SharedPreferences;
 import android.os.Bundle;
 import android.os.Handler;
 import android.os.RemoteException;
+import android.os.SystemProperties;
 import android.os.UserHandle;
 import android.provider.Settings;
 import android.util.ArraySet;
@@ -103,6 +104,7 @@ public class ButtonSettings extends SettingsPreferenceFragment
             "torch_long_press_power_gesture";
     private static final String KEY_TORCH_LONG_PRESS_POWER_TIMEOUT =
             "torch_long_press_power_timeout";
+    private static final String KEY_EDGE_GESTURE = "edge_gesture";
     private static final String KEY_CLICK_PARTIAL_SCREENSHOT =
             "click_partial_screenshot";
     private static final String KEY_SWAP_CAPACITIVE_KEYS = "swap_capacitive_keys";
@@ -118,6 +120,7 @@ public class ButtonSettings extends SettingsPreferenceFragment
     private static final String CATEGORY_BACKLIGHT = "key_backlight";
     private static final String CATEGORY_NAVBAR = "navigation_bar_category";
     private static final String CATEGORY_EXTRAS = "extras_category";
+    private static final String PRODUCT = SystemProperties.get("ro.build.product", "");

     private ListPreference mBackLongPressAction;
     private ListPreference mHomeLongPressAction;
@@ -147,6 +150,7 @@ public class ButtonSettings extends SettingsPreferenceFragment
     private SwitchPreference mHomeAnswerCall;
     private SwitchPreference mTorchLongPressPowerGesture;
     private ListPreference mTorchLongPressPowerTimeout;
+    private SwitchPreference mEdgeGesture;
     private SwitchPreference mSwapCapacitiveKeys;

     private PreferenceCategory mNavigationPreferencesCat;
@@ -470,6 +474,17 @@ public class ButtonSettings extends SettingsPreferenceFragment
             mSwapCapacitiveKeys.setDependency(KEY_DISABLE_NAV_KEYS);
         }

+        // Edge gesture controller
+        mEdgeGesture = findPreference(KEY_EDGE_GESTURE);
+
+        if (mEdgeGesture != null) {
+            if (PRODUCT.equals("libra")) {
+                mEdgeGesture.setChecked(SystemProperties.getBoolean("persist.vendor.edge_touch_mode", false));
+            } else {
+                extrasCategory.removePreference(mEdgeGesture);
+            }
+        }
+
         // Override key actions on Go devices in order to hide any unsupported features
         if (ActivityManager.isLowRamDeviceStatic()) {
             String[] actionEntriesGo = res.getStringArray(R.array.hardware_keys_action_entries_go);
@@ -552,6 +567,13 @@ public class ButtonSettings extends SettingsPreferenceFragment
                 (incallHomeBehavior == LineageSettings.Secure.RING_HOME_BUTTON_BEHAVIOR_ANSWER);
             mHomeAnswerCall.setChecked(homeButtonAnswersCall);
         }
+
+        // Get edge gesture status.
+        if (mEdgeGesture != null) {
+            if (PRODUCT.equals("libra")) {
+                mEdgeGesture.setChecked(SystemProperties.getBoolean("persist.vendor.edge_touch_mode", false));
+            }
+        }
     }

     private ListPreference initList(String key, Action value) {
@@ -837,6 +859,11 @@ public class ButtonSettings extends SettingsPreferenceFragment
         } else if (preference == mHomeAnswerCall) {
             handleToggleHomeButtonAnswersCallPreferenceClick();
             return true;
+        } else if (preference == mEdgeGesture) {
+            if (PRODUCT.equals("libra")) {
+                SystemProperties.set("persist.vendor.edge_touch_mode", mEdgeGesture.isChecked() ? "true" : "false");
+            }
+            return true;
         }

         return super.onPreferenceTreeClick(preference);

diff --git a/src/org/lineageos/lineageparts/power/PerfProfileSettings.java b/src/org/lineageos/lineageparts/power/PerfProfileSettings.java
index 949562f..d3ba3d6 100644
--- a/src/org/lineageos/lineageparts/power/PerfProfileSettings.java
+++ b/src/org/lineageos/lineageparts/power/PerfProfileSettings.java
@@ -28,6 +28,7 @@ import android.graphics.drawable.AnimatedVectorDrawable;
 import android.net.Uri;
 import android.os.Bundle;
 import android.os.PowerManager;
+import android.os.SystemProperties;
 import android.provider.Settings.Global;
 import android.util.ArraySet;
 import android.util.TypedValue;
@@ -65,9 +66,13 @@ public class PerfProfileSettings extends SettingsPreferenceFragment
     private static final String KEY_AUTO_POWER_SAVE  = "auto_power_save";
     private static final String KEY_POWER_SAVE       = "power_save";
     private static final String KEY_PERF_SEEKBAR     = "perf_seekbar";
+    private static final String KEY_CPU_OVERCLOCK_CATEGORY = "cpu_overclock_category";
+    private static final String KEY_CPU_OVERCLOCK    = "cpu_overclock";
+    private static final String PRODUCT = SystemProperties.get("ro.build.product", "");

     private ListPreference mAutoPowerSavePref;
     private SwitchPreference   mPowerSavePref;
+    private SwitchPreference   mCPUOverclockPref;

     private SeekBarPreference        mPerfSeekBar;
     private StopMotionVectorDrawable mPerfDrawable;
@@ -96,6 +101,7 @@ public class PerfProfileSettings extends SettingsPreferenceFragment
         mPerfSeekBar = findPreference(KEY_PERF_SEEKBAR);
         mAutoPowerSavePref = findPreference(KEY_AUTO_POWER_SAVE);
         mPowerSavePref = findPreference(KEY_POWER_SAVE);
+        mCPUOverclockPref = findPreference(KEY_CPU_OVERCLOCK);

         mPowerManager = getActivity().getSystemService(PowerManager.class);
         mPerf = PerformanceManager.getInstance(getActivity());
@@ -126,6 +132,13 @@ public class PerfProfileSettings extends SettingsPreferenceFragment
         updateAutoPowerSaveValue();
         mAutoPowerSavePref.setOnPreferenceChangeListener(this);
         mPowerSavePref.setOnPreferenceChangeListener(this);
+
+        if (PRODUCT.equals("aqua") || PRODUCT.equals("libra")) {
+            mCPUOverclockPref.setChecked(SystemProperties.getBoolean("persist.sys.cpu_overclock", false));
+            mCPUOverclockPref.setOnPreferenceChangeListener(this);
+        } else {
+            removePreference(KEY_CPU_OVERCLOCK_CATEGORY);
+        }
     }


@@ -212,6 +225,12 @@ public class PerfProfileSettings extends SettingsPreferenceFragment
             getActivity().registerReceiver(mPowerSaveReceiver,
                     new IntentFilter(PowerManager.ACTION_POWER_SAVE_MODE_CHANGING));
         }
+
+        if (mCPUOverclockPref != null) {
+            if (PRODUCT.equals("aqua") || PRODUCT.equals("libra")) {
+                mCPUOverclockPref.setChecked(SystemProperties.getBoolean("persist.sys.cpu_overclock", false));
+            }
+        }
     }

     @Override
@@ -246,6 +265,10 @@ public class PerfProfileSettings extends SettingsPreferenceFragment
             final int level = Integer.parseInt((String) newValue);
             Global.putInt(getContentResolver(), Global.LOW_POWER_MODE_TRIGGER_LEVEL, level);
             updateAutoPowerSaveSummary(level);
+        } else if (preference == mCPUOverclockPref) {
+            if (PRODUCT.equals("aqua") || PRODUCT.equals("libra")) {
+                SystemProperties.set("persist.sys.cpu_overclock", mCPUOverclockPref.isChecked() ? "false" : "true");
+            }
         }
         return true;
     }

project packages/apps/PermissionController/
diff --git a/src/com/android/permissioncontroller/permission/utils/Utils.java b/src/com/android/permissioncontroller/permission/utils/Utils.java
index 65fdd59..fdd71e2 100644
--- a/src/com/android/permissioncontroller/permission/utils/Utils.java
+++ b/src/com/android/permissioncontroller/permission/utils/Utils.java
@@ -23,6 +23,7 @@ import static android.Manifest.permission_group.CALENDAR;
 import static android.Manifest.permission_group.CALL_LOG;
 import static android.Manifest.permission_group.CAMERA;
 import static android.Manifest.permission_group.CONTACTS;
+import static android.Manifest.permission_group.FAKE_PACKAGE;
 import static android.Manifest.permission_group.LOCATION;
 import static android.Manifest.permission_group.MICROPHONE;
 import static android.Manifest.permission_group.PHONE;
@@ -209,6 +210,8 @@ public final class Utils {

         PLATFORM_PERMISSIONS.put(Manifest.permission.BODY_SENSORS, SENSORS);

+        PLATFORM_PERMISSIONS.put(Manifest.permission.FAKE_PACKAGE_SIGNATURE, FAKE_PACKAGE);
+
         PLATFORM_PERMISSION_GROUPS = new ArrayMap<>();
         int numPlatformPermissions = PLATFORM_PERMISSIONS.size();
         for (int i = 0; i < numPlatformPermissions; i++) {

project packages/apps/Settings/
diff --git a/res/values-zh-rCN/cm_strings.xml b/res/values-zh-rCN/cm_strings.xml
index 2798a62..d7aa8ee 100644
--- a/res/values-zh-rCN/cm_strings.xml
+++ b/res/values-zh-rCN/cm_strings.xml
@@ -96,4 +96,7 @@
     <string name="wifi_setup_wizard_title">选择 Wi\u2011Fi 网络</string>
     <string name="advanced_notifications_title">高级通知设置</string>
     <string name="peak_refresh_rate_summary_custom">自动将某些内容的刷新频率从 60 Hz 调高到 %1$d Hz。这么做会增加耗电量。</string>
+    <string name="screen_refresh_rate">屏幕刷新率</string>
+    <string name="screen_refresh_rate_title">屏幕刷新率</string>
+    <string name="keywords_screen_refresh_rate">屏幕，刷新率</string>
 </resources>

diff --git a/res/values-zh-rHK/cm_strings.xml b/res/values-zh-rHK/cm_strings.xml
index 45b2af1..0b870f5 100644
--- a/res/values-zh-rHK/cm_strings.xml
+++ b/res/values-zh-rHK/cm_strings.xml
@@ -31,4 +31,7 @@
     <string name="wake_when_plugged_or_unplugged_title">當接上喚醒</string>
     <string name="wake_when_plugged_or_unplugged_summary">當連接或拔除電源時開啓螢幕</string>
     <string name="peak_refresh_rate_summary_custom">自動將部分內容的重新整理頻率，從 60 Hz 提升到 %1$d Hz。此功能會增加耗電量。</string>
+    <string name="screen_refresh_rate">螢幕更新率</string>
+    <string name="screen_refresh_rate_title">螢幕更新率</string>
+    <string name="keywords_screen_refresh_rate">螢幕，更新率</string>
 </resources>

diff --git a/res/values-zh-rTW/cm_strings.xml b/res/values-zh-rTW/cm_strings.xml
index bf1ab75..1b564a8 100644
--- a/res/values-zh-rTW/cm_strings.xml
+++ b/res/values-zh-rTW/cm_strings.xml
@@ -95,4 +95,7 @@
     <string name="wifi_setup_wizard_title">選取 Wi-Fi 網路</string>
     <string name="advanced_notifications_title">進階通知設定</string>
     <string name="peak_refresh_rate_summary_custom">自動將某些內容的重新整理頻率從 60 調高到 %1$d 赫茲 (Hz)。請注意，開啟這項設定會增加電池用量。</string>
+    <string name="screen_refresh_rate">螢幕更新率</string>
+    <string name="screen_refresh_rate_title">螢幕更新率</string>
+    <string name="keywords_screen_refresh_rate">螢幕，更新率</string>
 </resources>

diff --git a/res/values/cm_arrays.xml b/res/values/cm_arrays.xml
index aeb0be1..fd9817b 100644
--- a/res/values/cm_arrays.xml
+++ b/res/values/cm_arrays.xml
@@ -49,4 +49,27 @@
     <!-- Override supported color mode labels.
          NOTE: This list must have the exact same items count as config_availableColorModes -->
     <string-array name="available_vendor_color_modes" translatable="false" />
+
+    <!-- Display settings. The minimum refresh rate of screen. These are shown in a list dialog. -->
+    <string-array name="screen_refresh_rate_entries">
+        <item>30 Hz</item>
+        <item>48 Hz</item>
+        <item>50 Hz</item>
+        <item>60 Hz</item>
+        <item>75 Hz</item>
+    </string-array>
+
+    <!-- Do not translate. -->
+    <string-array name="screen_refresh_rate_values" translatable="false">
+        <!-- Do not translate. -->
+        <item>30</item>
+        <!-- Do not translate. -->
+        <item>48</item>
+        <!-- Do not translate. -->
+        <item>50</item>
+        <!-- Do not translate. -->
+        <item>60</item>
+        <!-- Do not translate. -->
+        <item>75</item>
+    </string-array>
 </resources>

diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index ab38dec..5bdeb87 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -175,4 +175,18 @@

     <!-- Display settings screen, peak refresh rate settings summary [CHAR LIMIT=NONE] -->
     <string name="peak_refresh_rate_summary_custom">Automatically raises the refresh rate from 60 to %1$d Hz for some content. Increases battery usage.</string>
+
+    <!-- Sound & display settings screen, setting option name to change screen refresh rate -->
+    <string name="screen_refresh_rate">Screen refresh rate</string>
+
+    <!-- Sound & display settings screen, setting option name to change screen refresh rate [CHAR LIMIT=30] -->
+    <string name="screen_refresh_rate_title">Screen refresh rate</string>
+
+    <!-- Sound & display settings screen, setting option summary to change screen refresh rate -->
+    <string name="screen_refresh_rate_summary">
+        <xliff:g id="refresh_rate_description">%1$s</xliff:g>
+    </string>
+
+    <!-- List of synonyms for the screen refresh rate setting, used to match in settings search [CHAR LIMIT=NONE] -->
+    <string name="keywords_screen_refresh_rate">screen, refresh rate</string>
 </resources>

diff --git a/res/xml/display_settings.xml b/res/xml/display_settings.xml
index 925cbb9..bfc3dc3 100644
--- a/res/xml/display_settings.xml
+++ b/res/xml/display_settings.xml
@@ -76,6 +76,14 @@
         android:entryValues="@array/screen_timeout_values"
         settings:keywords="@string/keywords_screen_timeout" />

+    <com.android.settings.display.RefreshRateListPreference
+        android:key="screen_refresh_rate"
+        android:title="@string/screen_refresh_rate"
+        android:summary="@string/summary_placeholder"
+        android:entries="@array/screen_refresh_rate_entries"
+        android:entryValues="@array/screen_refresh_rate_values"
+        settings:keywords="@string/keywords_screen_refresh_rate" />
+
     <Preference
         android:key="adaptive_sleep_entry"
         android:title="@string/adaptive_sleep_title"

diff --git a/src/com/android/settings/DisplaySettings.java b/src/com/android/settings/DisplaySettings.java
index 20a33f0..771f8e9 100644
--- a/src/com/android/settings/DisplaySettings.java
+++ b/src/com/android/settings/DisplaySettings.java
@@ -31,6 +31,7 @@ import com.android.settings.display.ShowOperatorNamePreferenceController;
 import com.android.settings.display.TapToWakePreferenceController;
 import com.android.settings.display.ThemePreferenceController;
 import com.android.settings.display.TimeoutPreferenceController;
+import com.android.settings.display.RefreshRatePreferenceController;
 import com.android.settings.display.VrDisplayPreferenceController;
 import com.android.settings.search.BaseSearchIndexProvider;
 import com.android.settingslib.core.AbstractPreferenceController;
@@ -49,6 +50,7 @@ public class DisplaySettings extends DashboardFragment {
     public static final String KEY_PROXIMITY_ON_WAKE = "proximity_on_wake";

     private static final String KEY_SCREEN_TIMEOUT = "screen_timeout";
+    private static final String KEY_SCREEN_REFRESH_RATE = "screen_refresh_rate";
     private static final String KEY_HIGH_TOUCH_SENSITIVITY = "high_touch_sensitivity_enable";

     @Override
@@ -91,6 +93,7 @@ public class DisplaySettings extends DashboardFragment {
         controllers.add(new ScreenSaverPreferenceController(context));
         controllers.add(new TapToWakePreferenceController(context));
         controllers.add(new TimeoutPreferenceController(context, KEY_SCREEN_TIMEOUT));
+        controllers.add(new RefreshRatePreferenceController(context, KEY_SCREEN_REFRESH_RATE));
         controllers.add(new VrDisplayPreferenceController(context));
         controllers.add(new ShowOperatorNamePreferenceController(context));
         controllers.add(new ThemePreferenceController(context));

diff --git a/src/com/android/settings/development/AdbRootPreferenceController.java b/src/com/android/settings/development/AdbRootPreferenceController.java
index 5f7d65d..8cdc24f 100644
--- a/src/com/android/settings/development/AdbRootPreferenceController.java
+++ b/src/com/android/settings/development/AdbRootPreferenceController.java
@@ -53,7 +53,7 @@ public class AdbRootPreferenceController extends DeveloperOptionsPreferenceContr

     @Override
     public boolean isAvailable() {
-        return Build.IS_DEBUGGABLE;
+        return true;
     }

     @Override

diff --git a/src/com/android/settings/display/RefreshRateListPreference.java b/src/com/android/settings/display/RefreshRateListPreference.java
new file mode 100644
index 0000000..8e04b21
--- /dev/null
+++ b/src/com/android/settings/display/RefreshRateListPreference.java
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.android.settings.display;
+
+import android.app.Dialog;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.util.AttributeSet;
+import android.view.View;
+
+import androidx.appcompat.app.AlertDialog.Builder;
+
+import com.android.settings.RestrictedListPreference;
+
+
+public class RefreshRateListPreference extends RestrictedListPreference {
+    private final CharSequence[] mInitialEntries;
+    private final CharSequence[] mInitialValues;
+
+    public RefreshRateListPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        mInitialEntries = getEntries();
+        mInitialValues = getEntryValues();
+    }
+
+    @Override
+    protected void onPrepareDialogBuilder(Builder builder, DialogInterface.OnClickListener listener) {
+        super.onPrepareDialogBuilder(builder, listener);
+        builder.setView(null);
+    }
+
+    @Override
+    protected void onDialogCreated(Dialog dialog) {
+        super.onDialogCreated(dialog);
+        dialog.create();
+    }
+}

diff --git a/src/com/android/settings/display/RefreshRatePreferenceController.java b/src/com/android/settings/display/RefreshRatePreferenceController.java
new file mode 100644
index 0000000..5513c71
--- /dev/null
+++ b/src/com/android/settings/display/RefreshRatePreferenceController.java
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.android.settings.display;
+
+import android.content.Context;
+import android.os.SystemProperties;
+import android.util.Log;
+
+import androidx.preference.Preference;
+
+import com.android.settings.R;
+import com.android.settings.core.PreferenceControllerMixin;
+import com.android.settingslib.core.AbstractPreferenceController;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+
+
+public class RefreshRatePreferenceController extends AbstractPreferenceController implements PreferenceControllerMixin, Preference.OnPreferenceChangeListener {
+    private static final String TAG = "RefreshRatePrefContr";
+    private final String mScreenRefreshRateKey;
+
+    public RefreshRatePreferenceController(Context context, String key) {
+        super(context);
+        mScreenRefreshRateKey = key;
+    }
+
+    @Override
+    public boolean isAvailable() {
+        final String product = SystemProperties.get("ro.build.product", "");
+        return product.equals("aqua") || product.equals("libra");
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return mScreenRefreshRateKey;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        final RefreshRateListPreference refreshrateListPreference = (RefreshRateListPreference) preference;
+        refreshrateListPreference.setValue(String.valueOf(getRefreshRate()));
+        updateRefreshRatePreferenceDescription(refreshrateListPreference, Integer.parseInt(refreshrateListPreference.getValue()));
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        final int value = Integer.parseInt((String) newValue);
+        setRefreshRate(value);
+        if (value != getRefreshRate()) {
+            Log.w(TAG, "cannot change screen refresh rate to unsupported value: " + String.valueOf(value));
+        }
+        updateRefreshRatePreferenceDescription((RefreshRateListPreference) preference, value);
+        return true;
+    }
+
+    public static CharSequence getRefreshRateDescription(int currentRefreshRate, CharSequence[] entries, CharSequence[] values) {
+        if (currentRefreshRate < 0 || entries == null || values == null || values.length != entries.length) {
+            return null;
+        }
+
+        for (int i = 0; i < values.length; i++) {
+            if (currentRefreshRate == Integer.parseInt(values[i].toString())) {
+                return entries[i];
+            }
+        }
+        return null;
+    }
+
+    private void updateRefreshRatePreferenceDescription(RefreshRateListPreference preference, int currentRefreshRate) {
+        final CharSequence[] entries = preference.getEntries();
+        final CharSequence[] values = preference.getEntryValues();
+        final CharSequence refreshrateDescription = getRefreshRateDescription(currentRefreshRate, entries, values);
+        final String summary = refreshrateDescription == null ? "" : mContext.getString(R.string.screen_refresh_rate_summary, refreshrateDescription);
+        preference.setSummary(summary);
+    }
+
+    private int getRefreshRate() {
+        File file = new File("/sys/class/graphics/fb0/dynamic_fps");
+        if (file.exists()) {
+            try {
+                FileReader filereader = new FileReader(file);
+                final int currentValue = (filereader.read() - '0') * 10 + (filereader.read() - '0');
+                filereader.close();
+                return currentValue;
+            } catch (Exception e) {
+                Log.e(TAG, "cannot get screen refresh rate", e);
+            }
+        } else {
+            Log.e(TAG, "File not found: /sys/class/graphics/fb0/dynamic_fps");
+        }
+        return -1;
+    }
+
+    private void setRefreshRate(int newValue) {
+        File file = new File("/sys/class/graphics/fb0/dynamic_fps");
+        if (file.exists()) {
+            try {
+                FileWriter filewriter = new FileWriter(file);
+                filewriter.write(String.valueOf(newValue));
+                filewriter.close();
+            } catch (Exception e) {
+                Log.e(TAG, "cannot set screen refresh rate", e);
+            }
+        } else {
+            Log.e(TAG, "File not found: /sys/class/graphics/fb0/dynamic_fps");
+        }
+    }
+}

project system/core/
diff --git a/adb/daemon/main.cpp b/adb/daemon/main.cpp
index 1e33128..6670fd7 100644
--- a/adb/daemon/main.cpp
+++ b/adb/daemon/main.cpp
@@ -73,17 +73,15 @@ static bool should_drop_privileges() {
     //
     // ro.secure:
     //   Drop privileges by default. Set to 1 on userdebug and user builds.
-    bool ro_secure = android::base::GetBoolProperty("ro.secure", true);
-    bool ro_debuggable = __android_log_is_debuggable();

     // Drop privileges if ro.secure is set...
-    bool drop = ro_secure;
+    bool drop = false;

     // ... except "adb root" lets you keep privileges in a debuggable build.
     std::string prop = android::base::GetProperty("service.adb.root", "");
     bool adb_root = (prop == "1");
     bool adb_unroot = (prop == "0");
-    if (ro_debuggable && adb_root) {
+    if (adb_root) {
         drop = false;
     }
     // ... and "adb unroot" lets you explicitly drop privileges.
@@ -119,12 +117,6 @@ static void drop_privileges(int server_port) {
     // Don't listen on a port (default 5037) if running in secure mode.
     // Don't run as root if running in secure mode.
     if (should_drop_privileges()) {
-        const bool should_drop_caps = !__android_log_is_debuggable();
-
-        if (should_drop_caps) {
-            minijail_use_caps(jail.get(), CAP_TO_MASK(CAP_SETUID) | CAP_TO_MASK(CAP_SETGID));
-        }
-
         minijail_change_gid(jail.get(), AID_SHELL);
         minijail_change_uid(jail.get(), AID_SHELL);
         // minijail_enter() will abort if any priv-dropping step fails.

diff --git a/adb/sockets.cpp b/adb/sockets.cpp
index 423af67..a70fe5f 100644
--- a/adb/sockets.cpp
+++ b/adb/sockets.cpp
@@ -416,7 +416,7 @@ asocket* create_local_service_socket(std::string_view name, atransport* transpor
     LOG(VERBOSE) << "LS(" << s->id << "): bound to '" << name << "' via " << fd_value;

 #if !ADB_HOST
-    if ((name.starts_with("root:") && getuid() != 0 && __android_log_is_debuggable()) ||
+    if ((name.starts_with("root:") && getuid() != 0) ||
         (name.starts_with("unroot:") && getuid() == 0) || name.starts_with("usb:") ||
         name.starts_with("tcpip:")) {
         D("LS(%d): enabling exit_on_close", s->id);

diff --git a/init/Android.bp b/init/Android.bp
index f292d68..b685e24 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -94,7 +94,7 @@ cc_defaults {
         "-Wthread-safety",
         "-DALLOW_FIRST_STAGE_CONSOLE=0",
         "-DALLOW_LOCAL_PROP_OVERRIDE=0",
-        "-DALLOW_PERMISSIVE_SELINUX=0",
+        "-DALLOW_PERMISSIVE_SELINUX=1",
         "-DREBOOT_BOOTLOADER_ON_PANIC=0",
         "-DWORLD_WRITABLE_KMSG=0",
         "-DDUMP_ON_UMOUNT_FAILURE=0",

diff --git a/init/Android.mk b/init/Android.mk
index 416b732..56a2478 100644
--- a/init/Android.mk
+++ b/init/Android.mk
@@ -18,7 +18,7 @@ else
 init_options += \
     -DALLOW_FIRST_STAGE_CONSOLE=0 \
     -DALLOW_LOCAL_PROP_OVERRIDE=0 \
-    -DALLOW_PERMISSIVE_SELINUX=0 \
+    -DALLOW_PERMISSIVE_SELINUX=1 \
     -DREBOOT_BOOTLOADER_ON_PANIC=0 \
     -DWORLD_WRITABLE_KMSG=0 \
     -DDUMP_ON_UMOUNT_FAILURE=0

diff --git a/libcutils/trace-dev.cpp b/libcutils/trace-dev.cpp
index 5a09a2d..080e671 100644
--- a/libcutils/trace-dev.cpp
+++ b/libcutils/trace-dev.cpp
@@ -24,23 +24,14 @@ static pthread_once_t atrace_once_control = PTHREAD_ONCE_INIT;
 // the Zygote process from tracing.
 void atrace_set_tracing_enabled(bool enabled)
 {
+    enabled = false;
     atomic_store_explicit(&atrace_is_enabled, enabled, memory_order_release);
     atrace_update_tags();
 }

 static void atrace_init_once()
 {
-    atrace_marker_fd = open("/sys/kernel/debug/tracing/trace_marker", O_WRONLY | O_CLOEXEC);
-    if (atrace_marker_fd == -1) {
-        atrace_marker_fd = open("/sys/kernel/tracing/trace_marker", O_WRONLY | O_CLOEXEC);
-    }
-
-    if (atrace_marker_fd == -1) {
-        ALOGE("Error opening trace file: %s (%d)", strerror(errno), errno);
-        atrace_enabled_tags = 0;
-    } else {
-      atrace_enabled_tags = atrace_get_property();
-    }
+    atrace_enabled_tags = 0;
 }

 static void atrace_seq_number_changed(uint32_t prev_seq_no, uint32_t seq_no) {

diff --git a/libcutils/trace-dev.inc b/libcutils/trace-dev.inc
index 6543426..a8fc338 100644
--- a/libcutils/trace-dev.inc
+++ b/libcutils/trace-dev.inc
@@ -52,7 +52,7 @@ atomic_bool              atrace_is_ready      = ATOMIC_VAR_INIT(false);
 int                      atrace_marker_fd     = -1;
 uint64_t                 atrace_enabled_tags  = ATRACE_TAG_NOT_READY;
 static bool              atrace_is_debuggable = false;
-static atomic_bool       atrace_is_enabled    = ATOMIC_VAR_INIT(true);
+static atomic_bool       atrace_is_enabled    = ATOMIC_VAR_INIT(false);
 static pthread_mutex_t   atrace_tags_mutex    = PTHREAD_MUTEX_INITIALIZER;

 /**
@@ -100,6 +100,7 @@ uint64_t atrace_get_enabled_tags()
 // is not set to '1'.
 void atrace_set_debuggable(bool debuggable)
 {
+    debuggable = false;
     atrace_is_debuggable = debuggable;
     atrace_update_tags();
 }

diff --git a/rootdir/init.rc b/rootdir/init.rc
index a9af0b0..429da4c 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -540,11 +540,6 @@ on post-fs-data
     # Make sure we have the device encryption key.
     installkey /data

-    # Start bootcharting as soon as possible after the data partition is
-    # mounted to collect more data.
-    mkdir /data/bootchart 0755 shell shell encryption=Require
-    bootchart start
-
     # Make sure that apexd is started in the default namespace
     enter_default_mount_ns

@@ -611,26 +606,13 @@ on post-fs-data
     mkdir /data/misc/audioserver 0700 audioserver audioserver
     mkdir /data/misc/cameraserver 0700 cameraserver cameraserver
     mkdir /data/misc/vold 0700 root root
-    mkdir /data/misc/boottrace 0771 system shell
-    mkdir /data/misc/update_engine 0700 root root
-    mkdir /data/misc/update_engine_log 02750 root log
-    mkdir /data/misc/trace 0700 root root
-    # create location to store surface and window trace files
-    mkdir /data/misc/wmtrace 0700 system system
     # profile file layout
     mkdir /data/misc/profiles 0771 system system
     mkdir /data/misc/profiles/cur 0771 system system
     mkdir /data/misc/profiles/ref 0770 system system
-    mkdir /data/misc/profman 0770 system shell
-    mkdir /data/misc/gcov 0770 root root
     mkdir /data/misc/installd 0700 root root
     mkdir /data/misc/apexdata 0711 root root
     mkdir /data/misc/apexrollback 0700 root root
-    mkdir /data/misc/snapshotctl_log 0755 root root
-    # create location to store pre-reboot information
-    mkdir /data/misc/prereboot 0700 system system
-
-    mkdir /data/preloads 0775 system system encryption=None

     mkdir /data/vendor 0771 root root encryption=Require
     mkdir /data/vendor_ce 0771 root root encryption=None
@@ -640,7 +622,6 @@ on post-fs-data
     # For security reasons, /data/local/tmp should always be empty.
     # Do not place files or directories in /data/local/tmp
     mkdir /data/local/tmp 0771 shell shell
-    mkdir /data/local/traces 0777 shell shell
     mkdir /data/data 0771 system system encryption=None
     mkdir /data/app-private 0771 system system encryption=Require
     mkdir /data/app-ephemeral 0771 system system encryption=Require
@@ -648,18 +629,9 @@ on post-fs-data
     mkdir /data/app-lib 0771 system system encryption=Require
     mkdir /data/app 0771 system system encryption=Require
     mkdir /data/property 0700 root root encryption=Require
-    mkdir /data/tombstones 0771 system system encryption=Require
-    mkdir /data/vendor/tombstones 0771 root root
-    mkdir /data/vendor/tombstones/wifi 0771 wifi wifi

     # create dalvik-cache, so as to enforce our permissions
     mkdir /data/dalvik-cache 0771 root root encryption=Require
-    # create the A/B OTA directory, so as to enforce our permissions
-    mkdir /data/ota 0771 root root encryption=Require
-
-    # create the OTA package directory. It will be accessed by GmsCore (cache
-    # group), update_engine and update_verifier.
-    mkdir /data/ota_package 0770 system cache encryption=Require

     # create resource-cache and double-check the perms
     mkdir /data/resource-cache 0771 system system encryption=Require
@@ -677,20 +649,11 @@ on post-fs-data
     # the following directory.
     mkdir /data/mediadrm 0770 mediadrm mediadrm encryption=Require

-    mkdir /data/anr 0775 system system encryption=Require
-
-    # NFC: create data/nfc for nv storage
-    mkdir /data/nfc 0770 nfc nfc encryption=Require
-    mkdir /data/nfc/param 0770 nfc nfc
-
     # Create all remaining /data root dirs so that they are made through init
     # and get proper encryption policy installed
     mkdir /data/backup 0700 system system encryption=Require
-    mkdir /data/ss 0700 system system encryption=Require

     mkdir /data/system 0775 system system encryption=Require
-    mkdir /data/system/dropbox 0700 system system
-    mkdir /data/system/heapdump 0700 system system
     mkdir /data/system/users 0775 system system

     mkdir /data/system_de 0770 system system encryption=None
@@ -728,17 +691,7 @@ on post-fs-data
     mkdir /data_mirror/cur_profiles 0700 root root
     mount none /data/misc/profiles/cur /data_mirror/cur_profiles bind rec

-    mkdir /data/cache 0770 system cache encryption=Require
-    mkdir /data/cache/recovery 0770 system cache
-    mkdir /data/cache/backup_stage 0700 system system
-    mkdir /data/cache/backup 0700 system system
-
-    # Delete these if need be, per b/139193659
-    mkdir /data/rollback 0700 system system encryption=DeleteIfNecessary
-    mkdir /data/rollback-observer 0700 system system encryption=DeleteIfNecessary
-
-    # Create root dir for Incremental Service
-    mkdir /data/incremental 0771 system system encryption=Require
+    symlink /cache /data/cache

     # Create directories for statsd
     mkdir /data/misc/stats-active-metric/ 0770 statsd system
@@ -978,12 +931,6 @@ on property:vold.decrypt=trigger_shutdown_framework
     class_reset_post_data core
     class_reset_post_data hal

-on property:sys.boot_completed=1
-    bootchart stop
-    # Setup per_boot directory so other .rc could start to use it on boot_completed
-    exec - system system -- /bin/rm -rf /data/per_boot
-    mkdir /data/per_boot 0700 system system encryption=Require key=per_boot_ref
-
 # system server cannot write to /proc/sys files,
 # and chown/chmod does not work for /proc/sys/ entries.
 # So proxy writes through init.
@@ -1043,13 +990,6 @@ service console /system/bin/sh
     seclabel u:r:shell:s0
     setenv HOSTNAME console

-on property:ro.debuggable=1
-    # Give writes to anyone for the trace folder on debug builds.
-    # The folder is used to store method traces.
-    chmod 0773 /data/misc/trace
-    # Give reads to anyone for the window trace folder on debug builds.
-    chmod 0775 /data/misc/wmtrace
-
 on init && property:ro.debuggable=1
     start console

project system/sepolicy/
diff --git a/Android.mk b/Android.mk
index 2245dff..a5bd6ac 100644
--- a/Android.mk
+++ b/Android.mk
@@ -85,10 +85,12 @@ ifndef BOARD_SEPOLICY_VERS
 BOARD_SEPOLICY_VERS := $(PLATFORM_SEPOLICY_VERSION)
 endif

-NEVERALLOW_ARG :=
+# Hacks to keep SELinux permissive in user builds
+SELINUX_IGNORE_NEVERALLOWS := true
+
 ifeq ($(SELINUX_IGNORE_NEVERALLOWS),true)
 ifeq ($(TARGET_BUILD_VARIANT),user)
-$(error SELINUX_IGNORE_NEVERALLOWS := true cannot be used in user builds)
+$(warning SELINUX_IGNORE_NEVERALLOWS := true cannot be used in user builds)
 endif
 $(warning Be careful when using the SELINUX_IGNORE_NEVERALLOWS flag. \
           It does not work in user builds and using it will \
@@ -1301,10 +1303,9 @@ $(built_sepolicy_neverallows)
 	$(hide) $(HOST_OUT_EXECUTABLES)/sepolicy-analyze $@.tmp permissive > $@.permissivedomains
 	$(hide) if [ "$(TARGET_BUILD_VARIANT)" = "user" -a -s $@.permissivedomains ]; then \
 		echo "==========" 1>&2; \
-		echo "ERROR: permissive domains not allowed in user builds" 1>&2; \
+		echo "WARNING: permissive domains not allowed in user builds" 1>&2; \
 		echo "List of invalid domains:" 1>&2; \
 		cat $@.permissivedomains 1>&2; \
-		exit 1; \
 		fi
 	$(hide) mv $@.tmp $@

@@ -1358,10 +1359,9 @@ $(LOCAL_BUILT_MODULE): $(sepolicy.recovery.conf) $(HOST_OUT_EXECUTABLES)/checkpo
 	$(hide) $(HOST_OUT_EXECUTABLES)/sepolicy-analyze $@.tmp permissive > $@.permissivedomains
 	$(hide) if [ "$(TARGET_BUILD_VARIANT)" = "user" -a -s $@.permissivedomains ]; then \
 		echo "==========" 1>&2; \
-		echo "ERROR: permissive domains not allowed in user builds" 1>&2; \
+		echo "WARNING: permissive domains not allowed in user builds" 1>&2; \
 		echo "List of invalid domains:" 1>&2; \
 		cat $@.permissivedomains 1>&2; \
-		exit 1; \
 		fi
 	$(hide) mv $@.tmp $@

project vendor/lineage/
diff --git a/config/common.mk b/config/common.mk
index 204da24..9e79c77 100644
--- a/config/common.mk
+++ b/config/common.mk
@@ -103,41 +103,21 @@ PRODUCT_RESTRICT_VENDOR_FILES := false
 # Bootanimation
 TARGET_SCREEN_WIDTH ?= 1080
 TARGET_SCREEN_HEIGHT ?= 1920
-PRODUCT_PACKAGES += \
-    bootanimation.zip

 # Lineage packages
 PRODUCT_PACKAGES += \
     LineageParts \
     LineageSettingsProvider \
     LineageSetupWizard \
-    Updater

 # Themes
 PRODUCT_PACKAGES += \
-    LineageThemesStub \
     ThemePicker

 # Config
 PRODUCT_PACKAGES += \
     SimpleDeviceConfig

-# Extra tools in Lineage
-PRODUCT_PACKAGES += \
-    7z \
-    bash \
-    curl \
-    getcap \
-    htop \
-    lib7z \
-    libsepol \
-    nano \
-    pigz \
-    setcap \
-    unrar \
-    vim \
-    zip
-
 # Filesystems tools
 PRODUCT_PACKAGES += \
     fsck.ntfs \
@@ -145,20 +125,6 @@ PRODUCT_PACKAGES += \
     mkfs.ntfs \
     mount.ntfs

-# Openssh
-PRODUCT_PACKAGES += \
-    scp \
-    sftp \
-    ssh \
-    sshd \
-    sshd_config \
-    ssh-keygen \
-    start-ssh
-
-# rsync
-PRODUCT_PACKAGES += \
-    rsync
-
 # Storage manager
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \
     ro.storage_manager.enabled=true

diff --git a/config/common_full.mk b/config/common_full.mk
index 91bf198..67bfa73 100644
--- a/config/common_full.mk
+++ b/config/common_full.mk
@@ -3,16 +3,6 @@ $(call inherit-product, vendor/lineage/config/common_mobile.mk)

 PRODUCT_SIZE := full

-# Include {Lato,Rubik} fonts
-$(call inherit-product-if-exists, external/google-fonts/lato/fonts.mk)
-$(call inherit-product-if-exists, external/google-fonts/rubik/fonts.mk)
-
-# Fonts
-PRODUCT_PACKAGES += \
-    fonts_customization.xml \
-    LineageLatoFont \
-    LineageRubikFont
-
 # Recorder
 PRODUCT_PACKAGES += \
     Recorder

diff --git a/config/common_mobile.mk b/config/common_mobile.mk
index 1f1c241..ef2eeb2 100644
--- a/config/common_mobile.mk
+++ b/config/common_mobile.mk
@@ -8,9 +8,7 @@ PRODUCT_PRODUCT_PROPERTIES += \

 # AOSP packages
 PRODUCT_PACKAGES += \
-    Email \
     ExactCalculator \
-    Exchange2

 # Lineage packages
 PRODUCT_PACKAGES += \
@@ -18,13 +16,6 @@ PRODUCT_PACKAGES += \
     Eleven \
     Etar \
     Jelly \
-    Profiles \
-    Seedvault
-
-ifneq ($(TARGET_EXCLUDES_AUDIOFX),true)
-PRODUCT_PACKAGES += \
-    AudioFX
-endif

 ifeq ($(PRODUCT_TYPE), go)
 PRODUCT_PACKAGES += \
@@ -40,20 +31,6 @@ PRODUCT_DEXPREOPT_SPEED_APPS += \
     TrebuchetQuickStep
 endif

-# Accents
-PRODUCT_PACKAGES += \
-    LineageBlackTheme \
-    LineageBlackAccent \
-    LineageBlueAccent \
-    LineageBrownAccent \
-    LineageCyanAccent \
-    LineageGreenAccent \
-    LineageOrangeAccent \
-    LineagePinkAccent \
-    LineagePurpleAccent \
-    LineageRedAccent \
-    LineageYellowAccent
-
 # Charger
 PRODUCT_PACKAGES += \
     charger_res_images
@@ -65,9 +42,10 @@ endif

 # Customizations
 PRODUCT_PACKAGES += \
-    IconShapeSquareOverlay \
     LineageNavigationBarNoHint \
-    NavigationBarMode2ButtonOverlay
+    NavigationBarMode2ButtonOverlay \
+    NavigationBarMode3ButtonOverlay \
+    NavigationBarModeGesturalOverlay

 # Media
 PRODUCT_SYSTEM_DEFAULT_PROPERTIES += \

diff --git a/config/telephony.mk b/config/telephony.mk
index 6adf48d..355cf57 100644
--- a/config/telephony.mk
+++ b/config/telephony.mk
@@ -9,7 +9,6 @@ PRODUCT_PACKAGES += \
 # Telephony packages
 PRODUCT_PACKAGES += \
     messaging \
-    Stk

 # Default ringtone
 PRODUCT_PRODUCT_PROPERTIES += \

diff --git a/overlay/common/frameworks/base/core/res/res/values/colors.xml b/overlay/common/frameworks/base/core/res/res/values/colors.xml
deleted file mode 100644
index 27fd690..0000000
--- a/overlay/common/frameworks/base/core/res/res/values/colors.xml
+++ /dev/null

diff --git a/prebuilt/common/etc/fonts_customization.xml b/prebuilt/common/etc/fonts_customization.xml
deleted file mode 100644
index 0c9a7fb..0000000
--- a/prebuilt/common/etc/fonts_customization.xml
+++ /dev/null

diff --git a/prebuilt/common/etc/init/lineage-ssh.rc b/prebuilt/common/etc/init/lineage-ssh.rc
deleted file mode 100644
index f3613c3..0000000
--- a/prebuilt/common/etc/init/lineage-ssh.rc
+++ /dev/null

diff --git a/prebuilt/common/etc/init/lineage-system.rc b/prebuilt/common/etc/init/lineage-system.rc
index f26a64b..deb4c99 100644
--- a/prebuilt/common/etc/init/lineage-system.rc
+++ b/prebuilt/common/etc/init/lineage-system.rc
@@ -8,9 +8,3 @@ on post-fs-data
     # Change permissions on fsck log so it can be added to the dropbox
     chown root log /dev/fscklogs/log
     chmod 0640 /dev/fscklogs/log
-
-# bugreport is triggered by holding down volume down, volume up and power
-service bugreport /system/bin/dumpstate -d -p -z
-    oneshot
-    disabled
-    keycodes 114 115 116

diff --git a/prebuilt/common/etc/init/lineage-updates.rc b/prebuilt/common/etc/init/lineage-updates.rc
deleted file mode 100644
index f1bed8d..0000000
--- a/prebuilt/common/etc/init/lineage-updates.rc
+++ /dev/null
